suite/performance/common.go

```diff
@@ -99,3 +108,246 @@ func recordResults(b ginkgo.Benchmarker, ID string, results []time.Duration) err
+			Int("Runs", len(jobRuns.Data)).
+			Msg("Total runs")
+		for _, jobDecodeData := range jobRuns.Data {
+			rqInts, err := actions.ExtractRequestIDFromJobRun(jobDecodeData)
+			if err != nil {
+				return err
+			}
+			rqID := common.Bytes2Hex(rqInts)
+			loc, _ := time.LoadLocation("UTC")
+			startTime := jobDecodeData.Attributes.CreatedAt.In(loc)
+			log.Debug().
+				Time("StartTime", startTime).
+				Str("RequestID", rqID).
+				Msg("Request found")
+			d := r.Get(rqID)
+			d.StartTime = startTime
+		}
+	}
+	return nil
+}
+
+func (r *PerfRequestIDTestResults) calculateLatencies(b ginkgo.Benchmarker) error {
+	var latencies []time.Duration
+	for rqID, testResult := range r.GetAll() {
+		latency := testResult.EndTime.Sub(testResult.StartTime)
+		log.Debug().
+			Str("RequestID", rqID).
+			Time("StartTime", testResult.StartTime).
+			Time("EndTime", testResult.EndTime).
+			Dur("Duration", latency).
+			Msg("Calculating latencies for request id")
+		if testResult.StartTime.IsZero() {
+			log.Warn().
+				Str("RequestID", rqID).
+				Msg("Start time zero")
+		}
+		if testResult.EndTime.IsZero() {
+			log.Warn().
+				Str("RequestID", rqID).
+				Msg("End time zero")
+		}
+		if latency.Seconds() < 0 {
+			log.Warn().
+				Str("RequestID", rqID).
+				Msg("Latency below zero")
+		} else {
+			latencies = append(latencies, latency)
+		}
+	}
+	if err := recordResults(b, "Request latency", latencies); err != nil {
+		return err
+	}
+	return nil
+}
+
+// NodeData common node data
+type NodeData interface {
+	GetJobID() string
+	GetProvingKeyHash() [32]byte
+}
+
+// RunlogNodeData node data required for runlog test
+type RunlogNodeData struct {
+	JobID string
+}
+
+// GetJobID gets internal job id
+func (n RunlogNodeData) GetJobID() string {
+	return n.JobID
+}
+
+// GetProvingKeyHash gets proving key hash for VRF
+func (n RunlogNodeData) GetProvingKeyHash() [32]byte {
+	return [32]byte{}
+}
+
+// VRFNodeData VRF node data
+type VRFNodeData struct {
+	ProvingKeyHash [32]byte
+	JobID          string
+}
+
+// GetJobID gets internal job id
+func (n VRFNodeData) GetJobID() string {
+	return n.JobID
+}
+
+// GetProvingKeyHash gets proving key hash for VRF
+func (n VRFNodeData) GetProvingKeyHash() [32]byte {
+	return n.ProvingKeyHash
+}
+
+// ContractsNodesJobsMap common contract to node to job id mapping for perf/soak tests
+type ContractsNodesJobsMap map[interface{}]map[client.Chainlink]NodeData
+
+// FromJobsChan fills ContractsNodesJobsMap from a chan used in parallel deployment
+func (c ContractsNodesJobsMap) FromJobsChan(jobsChan chan ContractsNodesJobsMap) {
+	for jobMap := range jobsChan {
+		for contractAddr, m := range jobMap {
+			if _, ok := c[contractAddr]; !ok {
+				c[contractAddr] = map[client.Chainlink]NodeData{}
+			}
+			for k, v := range m {
+				c[contractAddr][k] = v
+			}
+		}
+	}
+}
+
+// LimitErrGroup implements the errgroup.Group interface, but limits goroutines to a
+// execute at a max throughput.
+type LimitErrGroup struct {
```

Comment:
This is more for my understanding: Why rate limit the errors?

Reply:
Just a generic goroutine group with a rate-limiting so we don't hit node hard